File: Práctica4/java/src/deepspace/EnemyStarShip.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 *
 * @author daniel
 */
public class EnemyStarShip {
    /**
     * Nombre de la nave
     */
    private String name;
    
    /**
     * Poder del armamento de la nave
     */
    private float ammoPower;
    
    /**
     * Poder del escudo de la nave
     */
    private float shieldPower;
    
    /**
     * Loot asociado a la nave
     */
    private Loot loot;
    
    /**
     * Daño asociado a la nave
     */
    private Damage damage;
    
    /**
     * Constructor
     * @param n nombre
     * @param a poder de armamento
     * @param s escudo de la nave
     * @param l loot
     * @param d daño
     */
    
    public EnemyStarShip(String n, float a, float s, Loot l, Damage d) {
        name = n;
        ammoPower = a;
        shieldPower = s;
        loot = l;
        damage = d;
    }
    
    /**
     * Constructor de copia
     * @param e estación a copiar
     */
    public EnemyStarShip(EnemyStarShip e){
        name = e.name;
        ammoPower = e.ammoPower;
        shieldPower = e.shieldPower;
        loot = e.loot;
        damage = e.damage;
    }
    
    // Getters
    
    /**
     * Getter de name
     * @return name
     */
    public String getName(){
        return name;
    }
    
    /**
     * Getter de ammoPower
     * @return ammoPower
     */
    public float getAmmoPower(){
        return ammoPower;
    }
    
    /**
     * Getter de shieldPower
     * @return shieldPower
     */
    public float getShieldPower(){
        return shieldPower;
    }
    
    /**
     * Getter de loot
     * @return loot
     */
    public Loot getLoot(){
        return loot;
    }
    
    /**
     * Getter de damage
     * @return damage
     */
    public Damage getDamage(){
        return damage;
    }
    
    /**
     * Consulta escudos
     * @return shieldPower 
     */
    public float protection(){
        return shieldPower;
    }
    
    /**
     * Consulta potencia de fuego
     * @return ammoPower
     */
    public float fire(){
        return ammoPower;
    }
    
    public ShotResult receiveShot(float shot){
        if (shieldPower >= shot)
            return ShotResult.RESIST;
        else
            return ShotResult.DONOTRESIST;     
    }
    
     /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        String message = "[EnemyStarShip] -> Name: " + name
                + ", ammoPower: " + ammoPower
                + ", shieldPower: " + shieldPower
                + ", Loot: " + loot.toString()
                + ", Damage: " + damage.toString();
        return message;
    }
    
    /**
     * To UI.
     */
    EnemyToUI getUIversion() {
        return new EnemyToUI(this);
    }
}

File: Práctica4/Práctica 4-20210523/lib/EnemyStarShip.rb
#encoding:utf-8

require_relative 'EnemyToUI'
require_relative 'Damage'
require_relative 'Loot'
require_relative 'ShotResult'

module Deepspace

  class EnemyStarShip

    # Constructor
    # @param _name [String] nombre de la estación
    # @param _ammoPower [Float] potencia de fuego
    # @param _shieldPower [Float] poder de escudo
    # @param _loot [Loot] loot
    # @param _damage [Damage] daño
    def initialize(_name, _ammoPower, _shieldPower, _loot, _damage)
      # @!attribute [String] nombre de la estación
      @name = _name

      # @!attribute [Float] potencia de fuego
      @ammoPower = _ammoPower

      # @!attribute [Float] poder de escudo
      @shieldPower = _shieldPower

      # @!attribute [Loot] loot
      @loot = _loot

      # @!attribute [Damage] damage
      @damage = _damage
    end

    # Constructor de copia
    # @param enemy [EnemyStarShip] instancia que va a ser copiada
    # @return [EnemyStarShip] la copia
    def self.newCopy(enemy)
      return new(enemy.name, enemy.ammoPower, enemy.shieldPower, enemy.loot, enemy.damage)
    end

    attr_reader :name, :ammoPower, :shieldPower, :loot, :damage

    # Devuelve la potencia de disparo
    # @return [Float] ammoPower
    def fire
      return @ammoPower
    end

    # Devuelve la potencia de escudo
    # @return [Float] shieldPower
    def protection
      return @shieldPower
    end

    # Resultado de recibir un disparo de la estación
    # @param shot [Float] poder del disparo
    # @return [ShotResult] si resiste o no
    def receiveShot(shot)
      if @shieldPower >= shot
        return ShotResult::RESIST
      else
        return ShotResult::DONOTRESIST
      end
    end

    # String representation, UI version
    # ==========================================================================

    # String representation of the object
    # @return [String] string representation
    def to_s
      getUIversion().to_s
    end

    # To UI
    def getUIversion
      return EnemyToUI.new(self)
    end
  end
end

